File: Network.R
Network <- setRefClass(
  "Network",
  fields=list(
    layers="list",
    loss="function",
    loss_prime="function"
  ),
  methods=list(
    add = function(layer) {
      layers <<- c(layers, layer)
    },
    use = function(loss, loss_prime) {
      loss <<- loss
      loss_prime <<- loss_prime
    },
    predict = function(input_data) {
      samples <- nrow(input_data)
      result <- matrix(NA, samples)
      for (i in seq.int(samples)) {
        output <- matrix(input_data[i,],1,ncol(input_data))
        for (layer in layers) {
          output <- layer$forward_propagation(output)
        }
        result[i] <- output
      }
      return(result)
    },
    fit = function(x_train, y_train, epochs, learning_rate) {
      samples <- nrow(x_train)
      for (i in seq.int(epochs)) {
        err <- 0
        
        for (j in seq.int(samples)) {
          output <- matrix(x_train[j, ], 1, ncol(x_train))
          
          for (layer in layers) {
            output <- layer$forward_propagation(output)
          } # endfor layers
          
          err <- err + loss(y_train[j, ], output)
          
          error <- loss_prime(y_train[j, ], output)
          
          for (layer in rev(layers)) {
            error <- layer$backward_propagation(error, learning_rate)
          } # endfor rev layers
        } # endfor j
        err <- err / samples
        if (i%%100==0) {
          cat(paste0("epoch ",i,"/",epochs,"\terror=",signif(err,8),"\n"))
        }
      } # endfor i
    } # endfun fit
  )
)

File: Network.py
import random


class Network:
    def __init__(self):
        self.layers = []
        self.loss = None
        self.loss_prime = None

    # add layer to network
    def add(self, layer):
        self.layers.append(layer)

    # set loss to use
    def use(self, loss, loss_prime):
        self.loss = loss
        self.loss_prime = loss_prime

    # predict output for given input
    def predict(self, input_data):
        # sample dimension first
        samples = len(input_data)
        result = []

        # run network over all samples
        for i in range(samples):
            # forward propagation
            output = input_data[i]
            for layer in self.layers:
                output = layer.forward_propagation(output)
            result.append(output)

        return result

    # train the network
    def fit(self, x_train, y_train, epochs, learning_rate):
        # sample dimension first
        samples = len(x_train)

        # training loop
        for i in range(epochs):
            err = 0
            for j in range(samples):
                # forward propagation
                output = x_train[j]
                for layer in self.layers:
                    output = layer.forward_propagation(output)

                # compute loss (for display purpose only)
                err += self.loss(y_train[j], output)

                # backward propagation
                error = self.loss_prime(y_train[j], output)
                for layer in reversed(self.layers):
                    error = layer.backward_propagation(error, learning_rate)

            # calculate average error on all samples
            err /= samples
            print('epoch %d/%d   error=%f' % (i + 1, epochs, err))

    def fits(self, x_train, y_train, epochs, learning_rate, batch_size):
        samples = len(x_train)
        for i in range(epochs):
            err = 0
            batch = random.sample(range(samples), batch_size)
            for j in batch:
                output = x_train[j]
                for layer in self.layers:
                    output = layer.forward_propagation(output)
                err += self.loss(y_train[j], output)
                error = self.loss_prime(y_train[j], output)
                for layer in reversed(self.layers):
                    error = layer.backward_propagation(error, learning_rate)
            err /= batch_size
            if i % 100 == 0:
                print('epoch %d/%d   error=%f' % (i + 1, epochs, err))
